core/container.go

```diff
@@ -948,15 +950,35 @@ func (container *Container) Exec(ctx context.Context, gw bkgw.Client, defaultPla
 		case secret.EnvName != "":
 			secretDest = secret.EnvName
 			secretOpts = append(secretOpts, llb.SecretAsEnv(true))
+			secretsToScrub.Envs = append(secretsToScrub.Envs, secret.EnvName)
+
+			// we sort to avoid non-deterministic order that would break caching
+			sort.SliceStable(secretsToScrub.Envs, func(i, j int) bool {
+				return secretsToScrub.Envs[i] < secretsToScrub.Envs[j]
+			})
 		case secret.MountPath != "":
 			secretDest = secret.MountPath
+			secretsToScrub.Files = append(secretsToScrub.Files, secret.MountPath)
+
+			// we sort to avoid non-deterministic order that would break caching
+			sort.SliceStable(secretsToScrub.Files, func(i, j int) bool {
+				return secretsToScrub.Files[i] < secretsToScrub.Files[j]
+			})
```

Comment:
We could do this sorting once below instead (around line 975 where we check if they're non-empty):

```go
sort.Strings(secretsToScrub.Envs)
sort.Strings(secretsToScrub.Files)
```

Reply:
I wasn't sure that `sort.Strings` would be stable, but it turns out ok. I'll do that.

Reply:
Done.